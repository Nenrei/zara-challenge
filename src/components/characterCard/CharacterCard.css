.char-card {
    width: 173px;
    height: 246px;
    cursor: pointer;
}

@keyframes animacionInfinita {
    0% {
        opacity: 0.1;
    }
    50% {
        opacity: 0.4;
    }
    100% {
        opacity: 0.1;
    }
}

.char-card--placeholder {
    background-color: var(--color-dark);
    opacity: 0.2;
}

.char-card__thumbnail {
    width: 173px;
    height: 173px;
}

.char-card__thumbnail img {
    width: 173px;
    height: 173px;
    object-fit: cover;
    object-position: left;
}

.char-card__splitter {
    width: 100%;
    height: 5px;
    background-color: var(--marver-red);
}

.char-card__name {
    width: 100%;
    height: 56px;
    color: white;
    font-size: 14px;
    line-height: 16px;
    text-transform: uppercase;

    padding: 16px 16px 24px 16px;

    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;

    position: relative;
    overflow: hidden;

    background: linear-gradient(
            var(--marver-red),
            var(--marver-red) 50%,
            var(--color-dark) 50%,
            var(--color-dark)
        )
        100% 100%;
    background-size: 100% 200%;
    transition: background 0.3s;
}

.char-card__name__fav {
    min-width: 12px;
    height: 12px;

    transition-duration: 0.3s;
}

.char-card__name__text {
    cursor: text;
}

@media (min-width: 540px) {
    .char-card:hover .char-card__name {
        background-position: 0 0;
    }
    .char-card:hover .char-card__name__fav.icon--heart {
        background-color: var(--color-clear);
    }
}
