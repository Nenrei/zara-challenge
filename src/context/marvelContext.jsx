import React, { createContext, useState } from 'react';

const MarvelContext = createContext({});

const MarvelContextProvider = ({ defaultValue, children }) => {
    const [favoriteCharacters, setFavoriteCharacters] = useState([]);

    const isFavoriteCharacter = (id) => {
        return favoriteCharacters.find((el) => el.id === id) != null;
    };

    let value = {
        states: {
            favorite: {
                favoriteCharacters,
                setFavoriteCharacters,
            },
        },
        actions: {
            isFavoriteCharacter,
        },
    };

    if (defaultValue) {
        value = { ...value, ...defaultValue };
    }

    return <MarvelContext.Provider value={value}>{children}</MarvelContext.Provider>;
};

const useMarvelContext = () => {
    const context = React.useContext(MarvelContext);
    if (context === undefined) {
        throw new Error('useMarvelContext must be used within a MarvelContextProvider');
    }
    return context;
};

export { MarvelContextProvider, useMarvelContext };
