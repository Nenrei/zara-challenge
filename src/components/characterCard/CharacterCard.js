import React from 'react';
import './CharacterCard.css';
import PropTypes from 'prop-types';
import FavoriteButton from '../favoriteButton/FavoriteButton';
import { Link } from 'react-router-dom';

const CharacterCard = ({ characterData }) => {
    return (
        <div className="char-card" role="card">
            <Link to={`/character/${characterData.id}`} className="char-card__thumbnail">
                <img
                    src={`${characterData.thumbnail.path}.${characterData.thumbnail.extension}`}
                    alt={characterData.name}
                />
            </Link>
            <div className="char-card__splitter"></div>
            <div className="char-card__name corner-triangle">
                <div className="char-card__name__text">{characterData.name}</div>
                <FavoriteButton characterData={characterData} boldIcon={true} />
            </div>
        </div>
    );
};

CharacterCard.propTypes = {
    characterData: PropTypes.shape({
        thumbnail: PropTypes.shape({
            path: PropTypes.string.isRequired,
            extension: PropTypes.string.isRequired,
        }),
        name: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
    }).isRequired,
};

export default CharacterCard;
