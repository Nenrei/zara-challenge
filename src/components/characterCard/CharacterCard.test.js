import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import CharacterCard from './CharacterCard';
import { MarvelContextProvider } from '../../context/marvelContext';

it('No fav character card', () => {
    const component = renderer.create(
        <MarvelContextProvider>
            <BrowserRouter>
                <CharacterCard
                    characterData={{
                        id: 1009610,
                        name: 'Spider-Man (Peter Parker)',
                        description:
                            'Bitten by a radioactive spider, high school student Peter Parker gained the speed, strength and powers of a spider. Adopting the name Spider-Man, Peter hoped to start a career using his new abilities. Taught that with great power comes great responsibility, Spidey has vowed to use his powers to help people.',
                        thumbnail: {
                            path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/50/526548a343e4b',
                            extension: 'jpg',
                        },
                    }}
                ></CharacterCard>
            </BrowserRouter>
        </MarvelContextProvider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
it('Fav character card', () => {
    const component2 = renderer.create(
        <MarvelContextProvider
            defaultValue={{
                states: {
                    favorite: {
                        favoriteCharacters: [
                            {
                                id: 1009610,
                                name: 'Spider-Man (Peter Parker)',
                                description:
                                    'Bitten by a radioactive spider, high school student Peter Parker gained the speed, strength and powers of a spider. Adopting the name Spider-Man, Peter hoped to start a career using his new abilities. Taught that with great power comes great responsibility, Spidey has vowed to use his powers to help people.',
                                thumbnail: {
                                    path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/50/526548a343e4b',
                                    extension: 'jpg',
                                },
                            },
                        ],
                    },
                },
            }}
        >
            <BrowserRouter>
                <CharacterCard
                    characterData={{
                        id: 1009610,
                        name: 'Spider-Man (Peter Parker)',
                        description:
                            'Bitten by a radioactive spider, high school student Peter Parker gained the speed, strength and powers of a spider. Adopting the name Spider-Man, Peter hoped to start a career using his new abilities. Taught that with great power comes great responsibility, Spidey has vowed to use his powers to help people.',
                        thumbnail: {
                            path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/50/526548a343e4b',
                            extension: 'jpg',
                        },
                    }}
                ></CharacterCard>
            </BrowserRouter>
        </MarvelContextProvider>,
    );
    let tree2 = component2.toJSON();
    expect(tree2).toMatchSnapshot();
});
