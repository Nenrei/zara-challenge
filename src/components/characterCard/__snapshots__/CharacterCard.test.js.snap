// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fav character card 1`] = `
<div
  className="char-card"
  onClick={[Function]}
>
  <div
    className="char-card__thumbnail"
  >
    <img
      alt="Spider-Man (Peter Parker)"
      src="http://i.annihil.us/u/prod/marvel/i/mg/3/50/526548a343e4b.jpg"
    />
  </div>
  <div
    className="char-card__splitter"
  />
  <div
    className="char-card__name corner-triangle"
  >
    <div
      className="char-card__name__text"
    >
      Spider-Man (Peter Parker)
    </div>
    <div
      className="char-card__name__fav icon icon--heart-line-bold"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`No fav character card 1`] = `
<div
  className="char-card"
  onClick={[Function]}
>
  <div
    className="char-card__thumbnail"
  >
    <img
      alt="Spider-Man (Peter Parker)"
      src="http://i.annihil.us/u/prod/marvel/i/mg/3/50/526548a343e4b.jpg"
    />
  </div>
  <div
    className="char-card__splitter"
  />
  <div
    className="char-card__name corner-triangle"
  >
    <div
      className="char-card__name__text"
    >
      Spider-Man (Peter Parker)
    </div>
    <div
      className="char-card__name__fav icon icon--heart-line-bold"
      onClick={[Function]}
    />
  </div>
</div>
`;
